// nd-table/bin/output-formatter/output-formatters.const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.outputFormatters = void 0;
var output_formatter_ascii_1 = require("./output-formatter-ascii");
var output_formatter_csv_1 = require("./output-formatter-csv");
var output_formatter_html_1 = require("./output-formatter-html");
var output_formatter_json_1 = require("./output-formatter-json");
var output_formatter_markdown_1 = require("./output-formatter-markdown");
var output_formatter_tsv_1 = require("./output-formatter-tsv");
var output_formatter_utf8_1 = require("./output-formatter-utf8");
exports.outputFormatters = {
    ascii: output_formatter_ascii_1.outputFormatterASCII,
    csv: output_formatter_csv_1.outputFormatterCSV,
    html: output_formatter_html_1.outputFormatterHTML,
    json: output_formatter_json_1.outputFormatterJSON,
    markdown: output_formatter_markdown_1.outputFormatterMarkdown,
    tsv: output_formatter_tsv_1.outputFormatterTSV,
    utf8: output_formatter_utf8_1.outputFormatterUTF8
};
